### ISSUE TITLE
.duplicating() freshly created ImageTextures leads to corruption/crashes

### ISSUE THREAD

**Issue by Macksaur** [godotengine/godot#54930]
### Godot version

3.4.stable

### System information

Windows 10, GLES3, NVIDIA 471.41

### Issue description

If an `ImageTexture` is `.duplicated()` after immediately being created and having it's data set, the duplicate will have corrupted texture data.

Also, this appears to corrupt the heap and crash the application if the scale is sufficiently large (>2 works).

If I create an `ImageTexture` and fill it, then duplicate it, I'd expect two copies of the final product without corruption. I'd rather not have to wait on texture data being filled before I'm allowed to duplicate objects.

### Steps to reproduce

Attaching the following script to a `TextureRect` that fills the screen produces corrupted graphics.

```
extends TextureRect

var main:ImageTexture
var other:ImageTexture

var test_image := true

var scale := 0.5 # displays corrupted memory
#var scale := 2.0 # outright crashes

func _bind(to_texture:ImageTexture, image:Image) -> void:
	image = image.duplicate() # optional, but the real project does similar
	image.resize(image.get_width()*scale, image.get_height()*scale, Image.INTERPOLATE_NEAREST)
	to_texture.set_data(image)

func _ready() -> void:
	var image := Image.new()

	if test_image:
		# test with an image
		image.load("res://icon.png")
	else:
		# or test with a solid fill
		image.create(32, 32, false, Image.FORMAT_RGBA8)
		image.fill(Color.blue)

	main = ImageTexture.new()
	main.create(image.get_width(), image.get_height(), Image.FORMAT_RGBA8, Texture.FLAG_FILTER)

	# if the texture has data set before duplicating, it becomes corrupted
	_bind(main, image)

	other = main.duplicate()

# swap between the two textures for debugging
var flicker := true
var flicker_time := 0.0
func _process(delta: float) -> void:
	flicker_time = move_toward(flicker_time, 0, delta)
	if flicker_time == 0:
		texture = main if flicker else other
		flicker_time = 0.5
		flicker = not flicker
```

**Comment by Macksaur** [comment]
Theory: Race condition in `Image.resize()`? 🤷‍♂️

**Comment by nobuyukinyuu** [comment]
Under `set_data()` it states that the image dimensions, format, and mipmaps need to match the existing texture's image configuration.  I'm assuming that when you resize the image, the imagetexture resource no longer has valid bounds compared to what's in memory (maybe increasing the scale when doing this clobbered memory), so setting `other` to a duplicate of that probably grabs some kind of invalid state.

**Comment by nobuyukinyuu** [comment]
Update 1 : Trying to fetch the size using `get_size()` of the respective image in the above code reports correctly after resizing, but the resource from `main.get_data()` does not reflect the new size, so indeed it seems using `set_data()` is not updating the resource, This is the source code for that method, and it doesn't seem to hold your hand at all with regards to what you feed it, except that it expects the image you set is not null:

```cpp
void ImageTexture::set_data(const Ref<Image> &p_image) {
	ERR_FAIL_COND_MSG(p_image.is_null(), "Invalid image");

	VisualServer::get_singleton()->texture_set_data(texture, p_image);

	_change_notify();
	emit_changed();

	alpha_cache.unref();
	image_stored = true;
}
```

This makes a call to the current rasterizer loaded for your project (GLES2 or GLES3 for godot 3 projects) and somewhere in that method is probably where the state gets corrupted.  If you want to resize an ImageTexture, you probably want to use `ImageTexture.set_size_override()` instead, or call `main.crate_from_image()` again after resizing before calling `set_data()`, or else your duplicate will be accessing uninitialized memory (or out of bounds memory). 

I have to assume this issue has existed for a while, since I found what may be a [similar problem](https://godotengine.org/qa/6006/error-texture_set_data-condition-p_image-empty-is-true) from the Godot Q&A going back to 2016.  Godot 4.0 seems to have changed the name of this method to `ImageTexture.update()` to better reflect what it's supposed to do (and what it's not expected to do), and it also has more bounds checks to make sure you don't blow stuff up using this method.

**Comment by Macksaur** [comment]
Thank you for you investigation! I understand how to avoid and fix this but I worry that merely documenting the heap corruption isn't enough. I lost an entire scene / unsaved code because this crashed in a tool script, that's not good!

If it's possible I'd like to see this become safer.

**Comment by nobuyukinyuu** [comment]
I'll try submitting a patch to 3.x to add more checks to the data sent to `set_data()` shortly.  :)

**Comment by Macksaur** [comment]
Do you know what code's responsible for the clobbering? I can't figure it out. But when the heap corruption sets in, I get notified on arbitrary allocations. In looking at `RasterizerStorageGLES3::texture_set_data`, it has some pathways that expect/adapt the image from different sizes.

**Comment by nobuyukinyuu** [comment]
Yes, I assume it's happening somewhere there, and that's as far as I got before I was in over my head, lol.  However that call seems very low-level and I believe is called all over the engine, so I won't be messing around in there.  It likely expects any other method which interacts with the user (ie:  you and me) to sanitize the inputs first so it can do what it's supposed to do as efficiently as possible.
